File: compiler/nodes/constraints/default.go
package constraints

import (
	"wipple/database"
	"wipple/nodes/types"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type DefaultConstraintNode struct {
	Parameter database.Node
	Value     database.Node
	Facts     *database.Facts
}

func (node *DefaultConstraintNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseDefaultConstraint(parser *syntax.Parser) (*DefaultConstraintNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("LeftParenthesis", syntax.TokenConfig{
		Reason: "between these parentheses",
	})
	if err != nil {
		return nil, err
	}

	parameterSpan := parser.Spanned()

	name, err := syntax.ParseTypeParameterName(parser)
	if err != nil {
		return nil, err
	}

	parameter := &types.TypeParameterNode{
		Name:  name,
		Infer: false,
		Value: nil,
		Facts: database.NewFacts(parameterSpan()),
	}

	_, err = parser.Token("AnnotateOperator")
	if err != nil {
		return nil, err
	}

	parser.Commit("in this type annotation")

	parser.ConsumeLineBreaks()

	value, err := types.ParseType(parser)
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightParenthesis")
	if err != nil {
		return nil, err
	}

	return &DefaultConstraintNode{
		Parameter: parameter,
		Value:     value,
		Facts:     database.NewFacts(span()),
	}, nil
}

func (node *DefaultConstraintNode) Visit(visitor *visit.Visitor) {
	visitConstraint(visitor, node)

	visitor.Visit(node.Parameter)
	visitor.Db.Graph.Edge(node.Parameter, node, "parameter")
	visitor.Visit(node.Value)
	visitor.Db.Graph.Edge(node.Value, node, "value")
	visitor.Constraint(typecheck.NewDefaultConstraint(node.Parameter, node.Value))
}

File: compiler/src/nodes/constraints/default.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenError},
    database::{Node, NodeRef},
    nodes::{parse_named_type_parameter, parse_type, visit_constraint},
    syntax::{ParseError, Parser, TokenKind},
    typecheck::DefaultConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct DefaultConstraintNode {
    pub parameter: NodeRef,
    pub value: NodeRef,
}

impl Node for DefaultConstraintNode {}

pub fn parse_default_constraint(
    parser: &mut Parser<'_>,
) -> Result<DefaultConstraintNode, ParseError> {
    parser.token_with_reason(TokenKind::LeftParenthesis, "between these parentheses")?;
    let parameter = parser.node(parse_named_type_parameter)?;
    parser.token(TokenKind::AnnotateOperator)?;
    parser.commit("in this type annotation");
    parser.consume_line_breaks();
    let value = parse_type(parser)?;
    parser.token(TokenKind::RightParenthesis)?;

    Ok(DefaultConstraintNode { parameter, value })
}

impl Visit for DefaultConstraintNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_constraint(node, visitor);

        visitor.visit(&self.parameter);

        visitor.visit(&self.value);
        visitor.edge(&self.value, node, "value");

        visitor.constraint(DefaultConstraint::new(
            self.parameter.clone(),
            self.value.clone(),
        ));
    }
}

impl Codegen for DefaultConstraintNode {
    fn codegen(&self, ctx: &mut CodegenCtx<'_>) -> Result<(), CodegenError> {
        Err(ctx.error())
    }
}
